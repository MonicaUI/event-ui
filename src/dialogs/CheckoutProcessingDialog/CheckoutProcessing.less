@import (reference) '~event-widgets/shared-styles/Variables.less';
@import (reference) '~event-widgets/shared-styles/Utility.less';

.icon{
  &:before{
    font-family: 'streamline';
  }
}

.icon {
  .icon();
}

.successCheck {
  composes: icon;
  font-size: 75px;
  position: absolute;
  left: 50%;
  top: 30%;
  color: @progressBarMessageColorSuccess;
  transform: translate( -50%, -50%);
  &:before {
    content: "\e943";
  }
}

.checkoutProcessingMessage {
  position: absolute;
  left: 50%;
  top: 70%;
  text-align: center;
  transform: translate( -50%, -50%);
  color: #6F7377;
}

.checkoutProcessingTitle {
  font-size: 40px;
}
.checkoutProcessingText {
  font-size: 18px;
  padding-bottom: 1rem;
}
.dialogContainer {
  height: 65vh;
  .dialogMediaQuery();
}
