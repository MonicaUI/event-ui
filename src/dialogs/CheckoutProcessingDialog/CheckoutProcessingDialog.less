@import "../styles/Dialog.less";

@dialogOverlayIndex: 1001;
@dialogIndex: 1002;
