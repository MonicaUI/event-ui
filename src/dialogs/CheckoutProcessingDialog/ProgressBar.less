@import "~event-widgets/shared-styles/Variables.less";
@import (reference) '~nucleus-core/less/MediaQuery.less';
@import "ProgressBarBase.less";
