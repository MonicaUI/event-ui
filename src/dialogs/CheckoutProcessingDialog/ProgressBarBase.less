@import (reference) '~nucleus-core/less/MediaQuery.less';

.container{
  display: inline-block;
  clear: both;
  text-align: center;
  position: absolute;
  left: 50%;
  top: 40%;
  transform: translate( -50%, -50%);
  width: @progressBarWidth;
  height: @progressBarHeight;
  * {
    box-sizing: border-box;
  }

  @media @medium {
    width: 80%;
  }
}

.indicator{
  background: @progressBarBackgroundColor;
  border-radius: @progressBarborderRadius;
  position: relative;
  overflow: hidden;
  height: @progressBarHeight;
}

.bar {
  background: @progressBarBackgroundColorFiller;
  animation: ProgressBar 1.1s infinite linear;
  display: inline-block;
  position: absolute;
  top: 0; left: 0; bottom: 0; right: 0;
  height: @progressBarHeight;
  -webkit-transition: width .1s ease-out;
    -moz-transition: width .1s ease-out;
         transition: width .1s ease-out;
}

.message {
  color: @progressBarMessageColorAlt;
  font-size: @progressBarMessageFontSize;
  font-weight: bold;
  font-family: @fontFamily;
  margin: auto;
  position: absolute;
  top: 0; left: 0; bottom: 0; right: 0;
  text-align: center;
  transition: all 1s;
  width: 100%;
}

.successMessage {
  color: @progressBarMessageColorSuccess;
}

@-webkit-keyframes ProgressBar {
  0% {
    background-position:0px;
  }
  100% {
    background-position:36px;
  }
}
@keyframes ProgressBar {
  0% {
    background-position:0px;
  }
  100% {
    background-position:36px;
  }
}
