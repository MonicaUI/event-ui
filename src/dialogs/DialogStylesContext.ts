import React from 'react';

/**
 * Default dialog styles context is an empty object. Each application should define the DialogStylesContext.Provider
 * at the top level of their apps (higher in the tree than any framework components) so that we can connect to
 * the data via the DialogStylesContext.Consumer.
 */
const DialogStylesContext = React.createContext({});
export default DialogStylesContext;
