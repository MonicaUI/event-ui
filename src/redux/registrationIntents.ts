/*
 * Declare both the type and value of each constant to be the same string literal so that flow doesn't infer the type of
 * each constant to be 'string' which is too general. We want it to be limited to a more specific type that can contain
 * only one specific string.
 */
export const NOT_REGISTERING = 'NOT_REGISTERING';
export const STARTING_REGISTRATION = 'STARTING_REGISTRATION';
export const REGISTERING = 'REGISTERING';
export const SAVING_REGISTRATION = 'SAVING_REGISTRATION';
export const CHECKING_OUT = 'CHECKING_OUT';
export const CHECKED_OUT = 'CHECKED_OUT';
export const CHECKED_OUT_PARTIALLY = 'CHECKED_OUT_PARTIALLY';
export const CANCELLING = 'CANCELLING';
export const FINALIZED_CANCEL_REGISTRATION = 'FINALIZED_CANCEL_REGISTRATION';
export const DECLINED_REGISTRATION = 'DECLINED_REGISTRATION';

export const RegistrationIntent =
  // @ts-expect-error ts-migrate(2362) FIXME: The left-hand side of an arithmetic operation must... Remove this comment to see the full error message
  typeof NOT_REGISTERING |
  // @ts-expect-error ts-migrate(2363) FIXME: The right-hand side of an arithmetic operation mus... Remove this comment to see the full error message
  typeof STARTING_REGISTRATION |
  // @ts-expect-error ts-migrate(2363) FIXME: The right-hand side of an arithmetic operation mus... Remove this comment to see the full error message
  typeof REGISTERING |
  // @ts-expect-error ts-migrate(2363) FIXME: The right-hand side of an arithmetic operation mus... Remove this comment to see the full error message
  typeof SAVING_REGISTRATION |
  // @ts-expect-error ts-migrate(2363) FIXME: The right-hand side of an arithmetic operation mus... Remove this comment to see the full error message
  typeof CHECKING_OUT |
  // @ts-expect-error ts-migrate(2363) FIXME: The right-hand side of an arithmetic operation mus... Remove this comment to see the full error message
  typeof CHECKED_OUT |
  // @ts-expect-error ts-migrate(2363) FIXME: The right-hand side of an arithmetic operation mus... Remove this comment to see the full error message
  typeof CHECKED_OUT_PARTIALLY |
  // @ts-expect-error ts-migrate(2363) FIXME: The right-hand side of an arithmetic operation mus... Remove this comment to see the full error message
  typeof CANCELLING |
  // @ts-expect-error ts-migrate(2363) FIXME: The right-hand side of an arithmetic operation mus... Remove this comment to see the full error message
  typeof FINALIZED_CANCEL_REGISTRATION |
  // @ts-expect-error ts-migrate(2363) FIXME: The right-hand side of an arithmetic operation mus... Remove this comment to see the full error message
  typeof DECLINED_REGISTRATION;
