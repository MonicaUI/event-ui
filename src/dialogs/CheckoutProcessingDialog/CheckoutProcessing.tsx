import React from 'react';
import ProgressBar from 'nucleus-core/progress/ProgressBar';
import ProgressBarStyle from './ProgressBar.less';
import CheckoutStyle from './CheckoutProcessing.less';

const CompletedGreenTick = () => <div className={CheckoutStyle.successCheck} />;

type CheckoutProcessingProps = {
  translate?: $TSFixMeFunction;
  percentComplete?: number;
  isSuccess?: boolean;
  messageOverride?: string;
  headerOverride?: string;
};

/*
 * This component contains a progressbar and a green checkmark and processing messages.
 * As green check mark and a 'processing complete message' show up when the progress bar hits 100%.
 * A 'processing message' shows up otherwise
 */
function CheckoutProcessing({
  translate,
  percentComplete,
  isSuccess,
  messageOverride,
  headerOverride
}: CheckoutProcessingProps): $TSFixMe {
  const processingHeader =
    percentComplete < 100
      ? translate('_checkoutProcessingHeader_resx')
      : translate('_checkoutProcessingCompleteHeader_resx');
  const processingMessage =
    percentComplete < 100
      ? translate('_checkoutProcessingText_resx')
      : translate('_checkoutProcessingCompleteMessage_resx');
  return (
    <div className={CheckoutStyle.dialogContainer}>
      {percentComplete === 100 ? <CompletedGreenTick /> : null}
      <ProgressBar classes={ProgressBarStyle} success={isSuccess} digit={percentComplete} total={100} />
      <div>
        <div className={CheckoutStyle.checkoutProcessingMessage}>
          <h2 className={CheckoutStyle.checkoutProcessingTitle}>{headerOverride || processingHeader}</h2>
          <p className={CheckoutStyle.checkoutProcessingText}>{messageOverride || processingMessage}</p>
        </div>
      </div>
    </div>
  );
}
CheckoutProcessing.displayName = 'CheckoutProcessing';

export default CheckoutProcessing;
