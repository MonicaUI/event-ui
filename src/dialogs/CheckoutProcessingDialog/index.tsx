import React from 'react';
import { connect } from 'react-redux';
import Dialog from 'nucleus-core/dialog/Dialog';
import CheckoutProcessing from './CheckoutProcessing';
import {
  CHECKING_OUT,
  CHECKED_OUT,
  CHECKED_OUT_PARTIALLY,
  STARTING_REGISTRATION
} from '../../redux/registrationIntents';
import DialogStylesContext from '../DialogStylesContext';

const popupConfig = {
  requestClose: () => {},
  dialogId: 'CheckoutProcessingDialog',
  isModal: true,
  header: <div />,
  isAnimatable: false
};

type Props = {
  isPending: boolean;
};

function CheckoutProcessingDialog({ isPending, ...others }: Props) {
  return (
    <DialogStylesContext.Consumer>
      {dialogStyles => (
        <Dialog
          isOpen={isPending}
          {...popupConfig}
          classes={{
            ...(dialogStyles as $TSFixMe).checkoutProcessing,
            transition: { ...(dialogStyles as $TSFixMe).transitions.up }
          }}
        >
          <CheckoutProcessing {...others} />
        </Dialog>
      )}
    </DialogStylesContext.Consumer>
  );
}

CheckoutProcessingDialog.displayName = 'CheckoutProcessingDialog';

export default connect((state: $TSFixMe) => {
  const {
    regCartStatus: { checkoutProgress, registrationIntent, startProgress },
    registrationForm: {
      regCart: { regDecline }
    }
  } = state;
  if (registrationIntent === STARTING_REGISTRATION && typeof startProgress === 'number') {
    /*
     * Hack to get quick support for async reg mod creation
     * Should get a real UI treatment
     */
    return {
      messageOverride: state.text.translate('EventGuestSide_StartRegModProcessingMessage__resx'),
      translate: state.text.translate,
      percentComplete: startProgress,
      isPending: true,
      isSuccess: false
    };
  }
  return {
    messageOverride: regDecline && state.text.translate('EventGuestSide_DeclineProcessingMessage__resx'),
    headerOverride: regDecline && state.text.translate('EventGuestSide_DeclineProcessingHeader__resx'),
    translate: state.text.translate,
    percentComplete: checkoutProgress,
    isPending: registrationIntent === CHECKING_OUT,
    isSuccess: registrationIntent === CHECKED_OUT || registrationIntent === CHECKED_OUT_PARTIALLY
  };
})(CheckoutProcessingDialog);
