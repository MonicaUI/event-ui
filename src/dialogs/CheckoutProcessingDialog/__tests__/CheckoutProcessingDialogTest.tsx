import { mount } from 'enzyme';
import { Provider } from 'react-redux';
import createStoreWithMiddleware from 'nucleus-guestside-site/src/redux/devCreateStoreWithMiddleware';
import { combineReducers } from 'redux';
import React from 'react';
import { wait } from '../../../testUtils';
import CheckoutProcessingDialog from '..';
import CheckoutProcessing from '../CheckoutProcessing';
import { CHECKING_OUT, CHECKED_OUT } from '../../../redux/registrationIntents';
import DialogStylesContext from '../../DialogStylesContext';

jest.mock('nucleus-core/containers/Transition');

const store = createStoreWithMiddleware(
  combineReducers({
    text: (x = {}) => x,
    regCartStatus: (x = {}, action = {}) => {
      if (action.type === 'SET_INTENT') {
        // @ts-expect-error ts-migrate(2698) FIXME: Spread types may only be created from object types... Remove this comment to see the full error message
        return { ...x, registrationIntent: action.val };
      }
      return x;
    },
    registrationForm: (x = {}) => x
  }),
  {
    text: {
      translate: (resx, tokens) => (tokens ? `${resx}:${JSON.stringify(tokens)}` : resx)
    },
    regCartStatus: {
      registrationIntent: CHECKING_OUT
    },
    registrationForm: {
      regCart: {}
    }
  }
);

const dialogStyles = { transitions: {}, base: {} };

describe('CheckoutProcessingDialog', () => {
  test('matches snapshot when dialog is shown', async () => {
    const dialog = mount(
      <Provider store={store}>
        <DialogStylesContext.Provider value={dialogStyles}>
          <CheckoutProcessingDialog />
        </DialogStylesContext.Provider>
      </Provider>
    );
    expect(dialog.find(CheckoutProcessing)).toMatchSnapshot('dialog shown');
  });

  test('matches snapshot when dialog is shown and then closed', async () => {
    const dialog = mount(
      <Provider store={store}>
        <DialogStylesContext.Provider value={dialogStyles}>
          <CheckoutProcessingDialog />
        </DialogStylesContext.Provider>
      </Provider>
    );
    expect(dialog.find(CheckoutProcessing)).toMatchSnapshot('dialog shown before closing');
    store.dispatch({
      type: 'SET_INTENT',
      val: CHECKED_OUT
    });
    wait(0);
    expect(dialog.find(CheckoutProcessing)).toMatchSnapshot('dialog closed');
  });

  test('matches snapshot when decline reg', async () => {
    const declineStore = createStoreWithMiddleware(
      combineReducers({
        text: (x = {}) => x,
        regCartStatus: (x = {}, action = {}) => {
          if (action.type === 'SET_INTENT') {
            // @ts-expect-error ts-migrate(2698) FIXME: Spread types may only be created from object types... Remove this comment to see the full error message
            return { ...x, registrationIntent: action.val };
          }
          return x;
        },
        registrationForm: (x = {}) => x
      }),
      {
        text: {
          translate: (resx, tokens) => (tokens ? `${resx}:${JSON.stringify(tokens)}` : resx)
        },
        regCartStatus: {
          registrationIntent: CHECKING_OUT
        },
        registrationForm: {
          regCart: { regDecline: true }
        }
      }
    );
    const dialog = mount(
      <Provider store={declineStore}>
        <DialogStylesContext.Provider value={dialogStyles}>
          <CheckoutProcessingDialog />
        </DialogStylesContext.Provider>
      </Provider>
    );
    expect(dialog.find('[children="EventGuestSide_DeclineProcessingHeader__resx"]')).toBeTruthy();
    expect(dialog.find('[children="EventGuestSide_DeclineProcessingMessage__resx"]')).toBeTruthy();
  });
});
