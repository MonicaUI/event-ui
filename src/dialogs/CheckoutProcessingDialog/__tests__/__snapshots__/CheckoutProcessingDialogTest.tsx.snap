// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`CheckoutProcessingDialog matches snapshot when dialog is shown and then closed: dialog closed 1`] = `
<div
  class="dialogContainer"
>
  <div
    class=""
  >
    <div
      class=""
    >
      <div
        class=""
      />
      <div
        class=""
      />
    </div>
  </div>
  <div>
    <div
      class="checkoutProcessingMessage"
    >
      <h2
        class="checkoutProcessingTitle"
      >
        _checkoutProcessingCompleteHeader_resx
      </h2>
      <p
        class="checkoutProcessingText"
      >
        _checkoutProcessingCompleteMessage_resx
      </p>
    </div>
  </div>
</div>
`;

exports[`CheckoutProcessingDialog matches snapshot when dialog is shown and then closed: dialog shown before closing 1`] = `
<div
  class="dialogContainer"
>
  <div
    class=""
  >
    <div
      class=""
    >
      <div
        class=""
      />
      <div
        class=""
      />
    </div>
  </div>
  <div>
    <div
      class="checkoutProcessingMessage"
    >
      <h2
        class="checkoutProcessingTitle"
      >
        _checkoutProcessingCompleteHeader_resx
      </h2>
      <p
        class="checkoutProcessingText"
      >
        _checkoutProcessingCompleteMessage_resx
      </p>
    </div>
  </div>
</div>
`;

exports[`CheckoutProcessingDialog matches snapshot when dialog is shown: dialog shown 1`] = `
<div
  class="dialogContainer"
>
  <div
    class=""
  >
    <div
      class=""
    >
      <div
        class=""
      />
      <div
        class=""
      />
    </div>
  </div>
  <div>
    <div
      class="checkoutProcessingMessage"
    >
      <h2
        class="checkoutProcessingTitle"
      >
        _checkoutProcessingCompleteHeader_resx
      </h2>
      <p
        class="checkoutProcessingText"
      >
        _checkoutProcessingCompleteMessage_resx
      </p>
    </div>
  </div>
</div>
`;
